diffrax/_solver/langevin_srk.py

```diff
@@ -0,0 +1,289 @@
+import abc
+from typing import Generic, TypeVar
+
+import equinox as eqx
+import jax.lax as lax
+import jax.numpy as jnp
+import jax.tree_util as jtu
+from jax import vmap
+from jax._src.tree_util import PyTreeDef
+from jaxtyping import Array, ArrayLike, PyTree
+
+from .._custom_types import (
+    AbstractBrownianIncrement,
+    BoolScalarLike,
+    DenseInfo,
+    RealScalarLike,
+)
+from .._local_interpolation import LocalLinearInterpolation
+from .._solution import RESULTS
+from .._term import _LangevinArgs, LangevinTerm, LangevinTuple, LangevinX
+from .base import AbstractItoSolver, AbstractStratonovichSolver
+
+
+_ErrorEstimate = TypeVar("_ErrorEstimate", None, LangevinTuple)
+
+
+class _AbstractCoeffs(eqx.Module):
+    @property
+    @abc.abstractmethod
+    def dtype(self):
+        raise NotImplementedError
+
+
+_Coeffs = TypeVar("_Coeffs", bound=_AbstractCoeffs)
+
+
+# TODO: I'm not sure if I can use the _Coeffs type here,
+# given that I do not use Generic[_Coeffs] in the class definition.
+# How should I work around this?
+class _SolverState(eqx.Module):
+    h: RealScalarLike
+    taylor_coeffs: PyTree[_Coeffs, "LangevinX"]
+    coeffs: _Coeffs  # type: ignore
+    rho: LangevinX
+    prev_f: LangevinX
+
+
+# CONCERNING COEFFICIENTS:
+# The coefficients used in a step of this SRK depend on
+# the time increment h, and the parameter gamma.
+# Assuming the modelled SDE stays the same (i.e. gamma is fixed),
+# then these coefficients must be recomputed each time h changes.
+# Furthermore, for very small h, directly computing the coefficients
+# via the function below can cause large floating point errors.
+# Hence, we pre-compute the Taylor expansion of the SRK coefficients
+# around h=0. Then we can compute the SRK coefficients either via
+# the Taylor expansion, or via direct computation.
+# In short the Taylor coefficients give a Taylor expansion with which
+# one can compute the SRK coefficients more precisely for a small h.
+
+
+class AbstractLangevinSRK(
+    AbstractStratonovichSolver[_SolverState],
+    AbstractItoSolver[_SolverState],
+    Generic[_Coeffs, _ErrorEstimate],
+):
```

Comment:
Can you add a docstring here?

FWIW if this is based on James' theory then I suspect we should use a more precise name than `AbstractLangevinSRK`. Call it `AbstractFosterLangevinSRK` or similar.

Reply:
Yes, I added a short docstring.

I aggre and I can try to convince him, but I doubt he will agree to this naming scheme. (we could do it without him knowing though :))